doc/rtd/topics/datasources/upcloud.rst

```diff
@@ -0,0 +1,33 @@
+.. _datasource_upcloud:
+
+UpCloud
+=============
+
+The `UpCloud`_ datasource consumes information from UpCloud's `metadata
+service`_. This metadata service serves information about the
+running server via HTTP over the address 169.254.169.254 available in every
+DHCP-configured interface. The metadata API endpoints are fully described in
+UpCloud documentation at
+`https://developers.upcloud.com/1.3/8-servers/#metadata-service
+<https://developers.upcloud.com/1.3/8-servers/#metadata-service>`_.
+
+Configuration
+-------------
+
+UpCloud's datasource can be configured as follows:
+
+  datasource:
+    UpCloud:
+      retries: 5
+      timeout: 2
+
```

Comment:
this largely seems like unnecessary configuration to me.
I believe that you, @ajmyyra, represent the cloud provider here.  Which would you rather have happen after the user sees a failure:
a.) they search the internet for docs and find that they can tweek 'retries' up to 10 and 'timeout' to 40, make that change and go on their "merry" way.
b.) they open an issue with cloud-init logs and you can fix your broken metadata server?

Its fine to keep it, but if it were me, I'd pick a sane value and ensure that my metadata servers responded inside that.